src/protocol/boolean/random_bits_generator.rs

```diff
@@ -0,0 +1,139 @@
+use super::solved_bits::{RandomBitsShare, SolvedBits};
+use crate::error::Error;
+use crate::ff::Field;
+use crate::protocol::context::SemiHonestContext;
+use crate::protocol::RecordId;
+use std::collections::HashMap;
+
+/// A struct that pre-generates and buffers random sharings of bits from the
+/// `SolvedBits` protocol. Any protocol who wish to use a random-bits can draw
+/// one by calling `take_one()`. It will call `SolvedBits` once the stock falls
+/// below `REFILL_THRESHOLD` until it fills up the empty slots.
+///
+/// Internally, it uses two pointers. `read_pointer` keeps track of the next
+/// item to return, and `write_pointer` points to the least recent empty slot.
+/// The pointers are `i8`. Incrementing them cause wrap around at the boundary,
+/// which makes the `buffer` hash map a ring buffer.
+#[allow(dead_code)]
+pub struct RandomBitsGenerator<'a, F: Field> {
+    context: SemiHonestContext<'a, F>,
+    buffer: HashMap<i8, RandomBitsShare<F>>,
+    counter: u32,
+    read_pointer: i8,
+    write_pointer: i8,
+    abort_count: u32,
+}
+
+#[allow(dead_code)]
+impl<'a, F: Field> RandomBitsGenerator<'a, F> {
+    const REFILL_THRESHOLD: u8 = 16;
+
+    pub fn new(context: SemiHonestContext<'a, F>) -> Self {
+        Self {
+            context,
+            buffer: HashMap::with_capacity(u8::MAX.into()),
+            counter: 0,
+            read_pointer: 0,
+            write_pointer: 0,
+            abort_count: 0,
+        }
+    }
+
+    /// Takes one `RandomBitsShare` instance out of the buffer. If the number
+    /// of buffered items fall below a threshold, it'll replenish.
+    pub async fn take_one(&mut self) -> Result<RandomBitsShare<F>, Error> {
+        self.refill_if_needed().await?;
+        let r = self
+            .buffer
+            .remove(&self.read_pointer)
+            .expect("unexpected buffer depletion");
+        self.read_pointer = self.read_pointer.wrapping_add(1);
+        Ok(r)
+    }
+
+    /// Checks if the stock is short, and replenish as needed. Specifically, it
+    /// will call `SolvedBits` once the stock falls below `REFILL_THRESHOLD`,
+    /// until the `write_pointer` is one place behind the `read_pointer`.
+    async fn refill_if_needed(&mut self) -> Result<(), Error> {
+        // intentional type coercing `i8` -> `u8`
+        #[allow(clippy::cast_sign_loss)]
+        if self.write_pointer.wrapping_sub(self.read_pointer) as u8 > Self::REFILL_THRESHOLD {
+            return Ok(());
+        }
+
+        while self.read_pointer.wrapping_sub(self.write_pointer) != 1 {
+            let record_id = RecordId::from(self.counter);
+            let r = SolvedBits::execute(self.context.clone(), record_id).await?;
```

Comment:
This will generate them sequentially, which will be needlessly slow. I'd suggest generating them in large batches. Basically, aim to fully replenish the buffer. Compute the number of empty slots, and invoke that many in parallel. If some fail, so be it. Just insert the successful ones into the ring buffer.

Reply:
I was thinking I have to filter the ones that failed, run it again, until all slots are filled. But ignoring those failed would be much simpler. Will update it.